### PR TITLE
InkWell can show focusColor optional on touch device

### LINKED ISSUE #70294: [mobile]  Selected DropdownMenuItem isn't highlighted on mobile only
## Steps to Reproduce
1. Run the code following on the mobile simulator or physical mobile device 

<details> 
<summary>code sample</summary> 

```dart
/// Flutter code sample for DropdownButton
// @dart = 2.9

// This sample shows a `DropdownButton` with a large arrow icon,
// purple text style, and bold purple underline, whose value is one of "One",
// "Two", "Free", or "Four".
//
// ![](https://flutter.github.io/assets-for-api-docs/assets/material/dropdown_button.png)

import 'package:flutter/material.dart';

void main() => runApp(MyApp());

/// This is the main application widget.
class MyApp extends StatelessWidget {
  static const String _title = 'Flutter Code Sample';

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: _title,
      home: Scaffold(
        appBar: AppBar(title: const Text(_title)),
        body: Center(
          child: MyStatefulWidget(),
        ),
      ),
    );
  }
}

/// This is the stateful widget that the main application instantiates.
class MyStatefulWidget extends StatefulWidget {
  MyStatefulWidget({Key key}) : super(key: key);

  @override
  _MyStatefulWidgetState createState() => _MyStatefulWidgetState();
}

/// This is the private State class that goes with MyStatefulWidget.
class _MyStatefulWidgetState extends State<MyStatefulWidget> {
  String dropdownValue = 'One';

  @override
  Widget build(BuildContext context) {
    return DropdownButton<String>(
      value: dropdownValue,
      icon: Icon(Icons.arrow_downward),
      iconSize: 24,
      elevation: 16,
      style: TextStyle(color: Colors.deepPurple),
      underline: Container(
        height: 2,
        color: Colors.deepPurpleAccent,
      ),
      onChanged: (String newValue) {
        setState(() {
          dropdownValue = newValue;
        });
      },
      items: <String>[
        'One',
        'Two',
        'Free',
        'Four',
        'Five',
        'Six',
        'Seven',
        'Eight',
        'Nine',
        'Ten',
        'Eleven',
        'Twelve',
        'Thirteen',
        'Fourteen',
        'Fifteen',
        'Sixteen',
        'Eighteen',
        'Nineteen',
        'Twenty',
        'Twenty One',
        'Twenty Two',
        'Twenty Three',
        'Twenty Four',
        'Twenty Five'
      ].map<DropdownMenuItem<String>>((String value) {
        return DropdownMenuItem<String>(
          value: value,
          child: Text(value),
        );
      }).toList(),
    );
  }
}

```

</details> 

2. Select an item and then close the dropdown menu, open it again, selected isn't highlighted

### Preview

| Mobile | Web & Desktop |
| --------------- | --------------- |
ezgif com-gif-maker (2) | ezgif com-gif-maker (3)

### Additional context:
1. Selected item is correctly highlighted on the web and desktop using the same code sample. 
2. Issue reproduced on `stable`, `beta`, `master`

<details> 
<summary>flutter doctor -v</summary> 

```bash
[✓] Flutter (Channel stable, 1.22.3, on Mac OS X 10.15.7 19H15, locale en-GB)
    • Flutter version 1.22.3 at /Users/tahatesser/Code/flutter_stable
    • Framework revision 8874f21e79 (13 days ago), 2020-10-29 14:14:35 -0700
    • Engine revision a1440ca392
    • Dart version 2.10.3

 
[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.2)
    • Android SDK at /Users/tahatesser/Code/sdk
    • Platform android-30, build-tools 30.0.2
    • ANDROID_HOME = /Users/tahatesser/Code/sdk
    • Java binary at: /Applications/Android
      Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6915495)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 12.2)
    • Xcode at /Volumes/SanDisk/Xcode-beta.app/Contents/Developer
    • Xcode 12.2, Build version 12B5044c
    • CocoaPods version 1.10.0

[!] Android Studio (version 4.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    ✗ Flutter plugin not installed; this adds Flutter specific functionality.
    ✗ Dart plugin not installed; this adds Dart specific functionality.
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6915495)

[✓] VS Code (version 1.51.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.16.0

[✓] Connected device (2 available)
    • RMX2001 (mobile)     • EUYTFEUSQSRGDA6D          • android-arm64 • Android 10
      (API 29)
    • Taha’s iPad (mobile) • 00008020-000255113EE8402E • ios           • iOS 14.2

! Doctor found issues in 1 category.
```

```bash
[✓] Flutter (Channel beta, 1.23.0-18.1.pre, on Mac OS X 10.15.7 19H15 x86_64, locale
    en-GB)
    • Flutter version 1.23.0-18.1.pre at /Users/tahatesser/Code/flutter_beta
    • Framework revision 198df796aa (4 weeks ago), 2020-10-15 12:04:33 -0700
    • Engine revision 1d12d82d9c
    • Dart version 2.11.0 (build 2.11.0-213.1.beta)

 
[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.2)
    • Android SDK at /Users/tahatesser/Code/sdk
    • Platform android-30, build-tools 30.0.2
    • ANDROID_HOME = /Users/tahatesser/Code/sdk
    • Java binary at: /Applications/Android
      Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6915495)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 12.2)
    • Xcode at /Volumes/SanDisk/Xcode-beta.app/Contents/Developer
    • Xcode 12.2, Build version 12B5044c
    • CocoaPods version 1.10.0

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6915495)

[✓] VS Code (version 1.51.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.16.0

[✓] Connected device (5 available)
    • RMX2001 (mobile)     • EUYTFEUSQSRGDA6D                     • android-arm64  •
      Android 10 (API 29)
    • Taha’s iPad (mobile) • 00008020-000255113EE8402E            • ios            •
      iOS 14.2
    • iPhone 11 (mobile)   • FF1F0A30-1571-44C9-B79D-E18CE0622B1B • ios            •
      com.apple.CoreSimulator.SimRuntime.iOS-14-2 (simulator)
    • Web Server (web)     • web-server                           • web-javascript •
      Flutter Tools
    • Chrome (web)         • chrome                               • web-javascript •
      Google Chrome 86.0.4240.193

• No issues found!
```

```bash
[✓] Flutter (Channel master, 1.24.0-8.0.pre.194, on Mac OS X 10.15.7 19H15
    darwin-x64, locale en-GB)
    • Flutter version 1.24.0-8.0.pre.194 at /Users/tahatesser/Code/flutter_master
    • Framework revision 018467cdb1 (8 hours ago), 2020-11-11 02:04:03 -0500
    • Engine revision 81f219c59c
    • Dart version 2.12.0 (build 2.12.0-31.0.dev)

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.2)
    • Android SDK at /Users/tahatesser/Code/sdk
    • Platform android-30, build-tools 30.0.2
    • ANDROID_HOME = /Users/tahatesser/Code/sdk
    • Java binary at: /Applications/Android
      Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6915495)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 12.2)
    • Xcode at /Volumes/SanDisk/Xcode-beta.app/Contents/Developer
    • Xcode 12.2, Build version 12B5044c
    • CocoaPods version 1.10.0

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6915495)

[✓] VS Code (version 1.51.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.16.0

[✓] Connected device (6 available)
    • RMX2001 (mobile)     • EUYTFEUSQSRGDA6D                     • android-arm64  •
      Android 10 (API 29)
    • Taha’s iPad (mobile) • 00008020-000255113EE8402E            • ios            •
      iOS 14.2
    • iPhone 11 (mobile)   • FF1F0A30-1571-44C9-B79D-E18CE0622B1B • ios            •
      com.apple.CoreSimulator.SimRuntime.iOS-14-2 (simulator)
    • macOS (desktop)      • macos                                • darwin-x64     •
      Mac OS X 10.15.7 19H15 darwin-x64
    • Web Server (web)     • web-server                           • web-javascript •
      Flutter Tools
    • Chrome (web)         • chrome                               • web-javascript •
      Google Chrome 86.0.4240.193

• No issues found!
```

</details>

### PR DESCRIPTION
**Pull request by xu-baolin**
## Description

Currently, the focus highlight can only be shown when `[FocusManager.highlightMode]` is `[FocusHighlightMode.traditional]`,
Sometimes, showing the focus highlight is useful for the touch device widget, such as highlight the selected item of `DropdownButton`, so I add a property `alwaysShowFocusColor` of `InkWell` to make it optinal.

I also improve some documentations about `focusColor` and `[FocusManager.highlightMode]` behavior.

## Related Issues

Fixes #70294 

## Tests

I added the following tests:

See files.

### CODE REVIEW

**Comment by gspencergoog** [comment]
Hi @xu-baolin,

Thanks for the PR, but I'd rather not add this parameter to `InkWell`, I think it's redundant.  You can already do what you want by setting [`FocusManager.instance.highlightStrategy`](https://master-api.flutter.dev/flutter/widgets/FocusManager/highlightStrategy.html) to `FocusHighlightStrategy.alwaysTraditional`, which will enable focus highlights to be shown even on a device in touch mode.

**Comment by xu-baolin** [comment]
But this doesn't seem to work  for this issue scenario. I just want to let some widget response instead of making all of them respond focus change.

**Comment by HansMuller** [comment]
I'd prefer to avoid adding one-off flags like alwaysShowFocusColor.

It's not possible to use the FocusManager, since it's a singleton and can't be used to override highlightStrategy within a subtree (i.e. for just one DropdownButton's menu).

I think what's really needed is a way to specify how menu item highlights are computed for the selected/focused/etc states. I think we could do that without hijacking the focus manager's highlightMode property, by exposing the per-item InkWell [overlayColor](https://api.flutter.dev/flutter/material/InkResponse/overlayColor.html) property in DropdownButton and by folding in the MaterialState.selected state, so overlayColor (`MaterialStateProperty<Color>`)values could select a color based on each item's disabled, hovered, pressed, focused, and selected state.

File: packages/flutter/lib/src/material/ink_well.dart [hunk 1]
```diff
@@ -930,17 +939,27 @@ class _InkResponseState extends State<_InkResponseStateWidget>
...
+    return widget.overlayColor?.resolve(selected) != null;
+  }
+
   void _updateFocusHighlights() {
     final bool showFocus;
+    final bool showSelect;
     switch (FocusManager.instance.highlightMode) {
       case FocusHighlightMode.touch:
         showFocus = false;
+        showSelect = true;
```
Comment (xu-baolin) [inline packages/flutter/lib/src/material/ink_well.dart:953]: For touch devices, do not display the focus color, try to display the selected color(setting by `overlayColor`) instead.

**Comment by HansMuller** [comment]
Why not "expose the per-item InkWell overlayColor property in DropdownButton" so that the color of menu items for all states could be specified with `DropdownButton.itemOverlayColor`?

File: packages/flutter/lib/src/material/ink_well.dart (hunk 1 above)
Comment (SaavyGirl) [reply packages/flutter/lib/src/material/ink_well.dart:965]: Looks good

Comment (SaavyGirl) [reply packages/flutter/lib/src/material/ink_well.dart:965]: Looks good

**Comment by xu-baolin** [comment]
> Why not "expose the per-item InkWell overlayColor property in DropdownButton" so that the color of menu items for all states could be specified with `DropdownButton.itemOverlayColor`?

Done.

**Comment by xu-baolin** [comment]
> How would a developer override the default overlayColor you've added?

Override the ThemeData for the subtree, the same as before this change.

**Comment by xu-baolin** [comment]
Hi guys, Any thoughts about this patch?

**Comment by HansMuller** [comment]
I appreciate the work you've put into this PR, and I apologize for only being a "drive by" reviewer. 

I think we're going to need to sort out a general solution per https://github.com/flutter/flutter/pull/70442#discussion_r528116468,  for all components that can take the focus when they're used in desktop applications.   I'm going to close this PR to give everyone a chance to think about how best to solve (and explain) a general solution.

Sorry about the long delay and for the effort you've put into this.

**Pull request closed without merging**